### PR TITLE
Expose a JDBC connection (session) via the ServiceHub for generic JDB…

### PR DESCRIPTION
**Pull request by josecoll**
…C usage.

### CODE REVIEW

**Comment by mikehearn** [comment]
Great. This should be documented in the right places in the docsite, so developers know they can use this feature and that it's available. And probably not just mentioned in passing, but integrated into the content so that it's clear in all the places where flows and data storage are discussed that we expect people to use this feature to solve various kinds of problems.

Can we expose some kind of SQL dialect identifier on the service hub too?

Perhaps it should just be called databaseConnection instead of jdbcConnection.

How would one write a flow that used a prepared statement with this API? Is it possible?

**Comment by josecoll** [comment]
Documentation noted.
My only concern right now is the potential destructive nature (inadvertently or maliciously) exposing this interface may cause. Without some kind of table/account management, we are effectively giving CorDapp developers carte blanche to perform any CRUD they like - on both our internal tables and their own contract state tables.
We really need some form of enforced permissioning - such that READ_ONLY access to internal tables is permissible (at minimum - there may be some internal information we do not wish to be queryable at all).
@rick-r3 thoughts?

**Comment by mikehearn** [comment]
Sure, we can log the connection in with a user that has limited permissions (in some future version of the node). We don't assume installed CorDapps are malicious though. If the app does something dumb, well, don't use that app. We should note in the docsite what the rules are around access to tables.

**Comment by gendal** [comment]
re the permissions on the connection...   how would using a different identity interact with transactionality?  More precisely, what do we want the user to _see_ when they query? (last clean state or the database's state with any current modifications that might be rolled back should the enclosing flow terminate and trigger a rollback?  Presumably the latter?)  And what should the lasting impact of any writes be?  In their own transactional context or committed/rolled back alongside (and atomically with) any transaction used by the flow to keep the checkpoints, artemis and other persistence stuff in sync?  The code says a transaction needs to be present... good.  So I guess my question nets down to: can you have two identities connecting to a database under the same transaction context in JDBC?  And, if not, can JDBC surface a "powerful" and a "weak" connection under the same ID?

**Comment by josecoll** [comment]
A transaction context is associated with a session (connection) which is associated with a single identity - so there is no possibility of clash / inconsistent state.
Typically a single generic account is used at the application layer to interact with a database.
However, I feel we need separate accounts for all CRUD (system level Node user) Vs Read-Only (CorDapp user interacting via JDBC).

**Comment by gendal** [comment]
Yes - but I was trying to understand the developer experience...   if they're interacting with our database in a flow, for example, through a connection we give them, won't they expect it to be coordinated in some way with the transaction that is maintaining the integrity of the node/checkpoint/flow/messaging, etc?

**Comment by josecoll** [comment]
By enforcing usage of a session within a transaction (the surrounding flow transaction boundary) we ensure that flow checkpointing takes into account any state retrieved at the time of checkpoint. 
Perhaps we need to validate this behaviour with some samples.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 1]
```diff
@@ -208,4 +209,13 @@ interface ServiceHub : ServicesForResolution {
      * @return A new [SignedTransaction] with the addition of the new signature.
      */
     fun addSignature(signedTransaction: SignedTransaction): SignedTransaction = addSignature(signedTransaction, legalIdentityKey)
+
+    /**
+     * Exposes a JDBC connection (session) object using the currently configured database.
+     * Applications can use this to execute arbitrary SQL queries (native, direct, prepared, callable)
+     * against its Node database tables (including custom contract tables defined by extending [Queryable]).
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: Re: Richard's questions. The API doc here definitely needs to discuss what can be seen when you query - for instance, anything done prior to the last checkpoint will show up in queries, other changes won't (or however db transactions are configured).

File: docs/source/api-persistence.rst [hunk 2]
```diff
@@ -92,3 +92,18 @@ Several examples of entities and mappings are provided in the codebase, includin
 
 .. literalinclude:: ../../finance/src/main/kotlin/net/corda/schemas/CashSchemaV1.kt
     :language: kotlin
+
+JDBC session
+------------
+CorDapp developer's may also interact directly with the underlying Node's database by using a standard
```
Comment (mikehearn) [inline docs/source/api-persistence.rst]: "CorDapp developer's" -> "Apps"  (apostrophe is for possession not plurals), lowercase node.

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import net.corda.core.serialization.SingletonSerializeAsToken
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.utilities.OpaqueBytes
+import net.corda.core.utilities.ProgressTracker
+import net.corda.core.utilities.loggerFor
+import net.corda.core.utilities.unwrap
+import net.corda.flows.*
+import java.util.*
+
+// START 1
```
Comment (mikehearn) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: Maybe we can move the doc start tag to just before line 33?

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        log.info("Service initialising ...")
+
+        log.info("Service initialisation completed successfully ...")
+    }
+
+    fun rebalanceCurrencyReserves() {
+        // search cash contract state table for all currencies and balances
+        val nativeQuery = """
+            select
+                sum(cashschema1_.pennies) as col_0_0_,
```
Comment (mikehearn) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: This looks like auto-generated SQL? Can we use something hand written and which looks like a join with a private dataset? That will be a clearer example.

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                vaultschem0_.output_index=cashschema1_.output_index
+                and vaultschem0_.transaction_id=cashschema1_.transaction_id
+                and vaultschem0_.state_status=?
+            group by
+                cashschema1_.ccy_code limit ?
+            order by
+                sum(cashschema1_.pennies)
+        """
+        val session = services.jdbcSession()
+        val prepStatement = session.prepareStatement(nativeQuery)
```
Comment (mikehearn) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: How do prepared statements interact with this API? Can they be used? Can they be stored and re-used across flows? The docs don't say. Maybe it's obvious to a more experienced JDBC user but it'd be nice to know the scope in which prepared statements live for.

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+// START 1
+@CordaService
+class CustomIssuerService(val services: PluginServiceHub) : SingletonSerializeAsToken() {
+
+    private companion object {
+        val log = loggerFor<CustomIssuerService>()
+    }
+
+    init {
```
Comment (mikehearn) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: This init block and the companion can be deleted, right?

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt [hunk 7]
```diff
@@ -0,0 +1,116 @@
+package net.corda.notarydemo
```
Comment (mikehearn) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: Where is this file used? I see it being added but I don't see where the docsite refers to it?

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt (hunk 1 above)
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: Please see updated discussion on https://r3-cev.atlassian.net/browse/CORPRIV-799

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt (hunk 7 above)
Comment (josecoll) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: Example fully updated.

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt (hunk 3 above)
Comment (josecoll) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: TBC

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt (hunk 4 above)
Comment (josecoll) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: TBC - using a simple native query on _contract_cash_states_ and _vault_states_ tables for now.

File: docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt (hunk 5 above)
Comment (josecoll) [inline docs/source/example-code/src/main/kotlin/net/corda/docs/CustomIssuerUsingJDBCTutorial.kt]: TBC

**Comment by josecoll** [comment]
@mikehearn @gendal 
Please see comments on https://r3-cev.atlassian.net/browse/CORPRIV-799 before continuing work on this PR.

**Comment by mikehearn** [comment]
There was no need for that story to be private so I moved it public here:

https://r3-cev.atlassian.net/browse/CORDA-488

WRT the comments - I didn't get from the Jira why we're now introducing a lambda requirement? What was wrong with the prior iteration of the API? Also - we do want to make Java 8 lambdas properly serializable in checkpoints, there's an open bug for that.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt (hunk 1 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: I don't see where that discussion is relevant to this PR comment. This API is a fine place to start - there is no need to add something in this PR for non-flow use cases, but it does need to discuss in the docsite+api docs how it interacts with flow transactions.

**Comment by josecoll** [comment]
@mikehearn please discuss with @rick-r3 on how you want me to proceed.
It is counter-productive to have 2 reviewers asking for different things.

**Comment by rick-r3** [comment]
The lambda (or command object if we don't want to do lambdas) is because we don't really want the connection to be in checkpoints since it's blacklisted by virtue of being closable.  We *could* allow that and do some work with serializers to support it or rely on the existing blacklisting logic to warn developers that it isn't allowed and they should do queries in a private method.  But confining their use to a defined scope avoids both of those but if we think it creates other problems then we can revert to either of the other two options.

**Comment by josecoll** [comment]
This PR is complete as per the original requirement.
Perhaps all it needs it a little more documentation.

**Comment by josecoll** [comment]
Ball back in your court @mikehearn

**Comment by mikehearn** [comment]
I figured you could just use a private method to control the scope of checkpointable things, or null it out: what stops someone doing a sendAndReceive in the lambda? I'm not sure it offers much protection to be honest, but whatever, it doesn't hurt much either.

**Pull request merged**